react-native-hub-app/src/checklist.tsx

```diff
@@ -0,0 +1,482 @@
+      return;
+    }
+    data.status = data.status += 1;
+    steps[index] = data;
+    this.setState({
+      steps: steps,
+    });
+  }
+
+  async generateIdentity(): Promise<Libp2pCryptoIdentity> {
+    let idStr = await AsyncStorage.getItem(IDENTITY_KEY);
+    if (idStr) {
+      return await Libp2pCryptoIdentity.fromString(idStr);
+    } else {
+      const id = await Libp2pCryptoIdentity.fromRandom();
+      idStr = id.toString();
+      await AsyncStorage.setItem(IDENTITY_KEY, idStr);
+      return id;
+    }
+  }
+
+  async getUserToken(database, id): Promise<string> {
+    let token;// = await AsyncStorage.getItem(TOKEN_KEY);
+    if (token) {
+      return token;
+    }
+    token = await database.getToken(id);
+    await AsyncStorage.setItem(TOKEN_KEY, token);
+    return token;
+  }
+
+  async getContext(): Promise<Context> {
+    // Pull the stored context to reuse if available && valid date
+    let contextStr = await AsyncStorage.getItem(CONTEXT_KEY);
+    if (contextStr) {
+      const ctxJson = JSON.parse(contextStr);
+      if (
+        ctxJson['x-textile-api-sig-msg'] && (Date.parse(ctxJson['x-textile-api-sig-msg'])) > (new Date()).getTime()) {
+        // Not expired
+        const ctx = Context.fromJSON(ctxJson, API_URL);
+        return ctx;
+      }
+    }
+  }
+
+  async getUserThread(db: Client, ctx: Context): Promise<ThreadID> {
+    let idStr = await AsyncStorage.getItem(USER_THREAD_ID);
+    if (idStr) {
+      /**
+       * Temporary hack to get ThreadID working in RN
+       */
+      const temp: ThreadID = ThreadID.fromString(idStr);
+      // @ts-ignore
+      const id = ThreadID.fromBytes(Buffer.from(temp.buf));
+      return id;
+    } else {
+      const id: ThreadID = ThreadID.fromRandom();
+      await AsyncStorage.setItem(USER_THREAD_ID, id.toString());
+
+      /**
+       * Each new ThreadID requires a `newDB` call.
+       */
+      await db.newDB(id, ctx.withThread(id))
+
+      /** 
+       * We add our first Collection to the DB for Astronauts.
+       */
+      await db.newCollection(id, 'Astronaut', astronautSchema);
+      return id;
+    }
+  }
+
+  async runStep(stepNumber: number) {
+    const steps = this.state.steps;
+    const data = steps[stepNumber];
+    try {
+      let {ctx, db, threadId} = this.state;
+      switch (stepNumber) {
+        case 0: {
+          let updatedState = {};
+          /**
+           * Create a new user Identity
+           * 
+           * The identity pk will be cached in AsyncStorage.
+           * On the next session, the pk identity will be reused
+           */
+          const id = await this.generateIdentity();
+          
+          /**
+           * Context contains the token and session information
+           * 
+           * If possible, we'll reuse an existing session. 
+           * If it doesn't exist or is expired, we'll create a new one.
+           */
+          const existingCtx = await this.getContext();
```

Comment:
This could probably be made into a single path by doing something like:

```
const ctx = await this.getContext() ?? new Context(API_URL);
db = new Client(ctx);
```

Then from then on, you can just reference the db's context directly

```
db.context.withBlah('something')
```